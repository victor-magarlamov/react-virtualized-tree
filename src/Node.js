export default class Node {
  static nodes = [];
  static FIRST_LEVEL = 1;

  item = null;

  isLeaf = false;

  open = false;

  level = Node.FIRST_LEVEL;

  constructor(item, isLeaf) {
    this.item = item;
    this.isLeaf = isLeaf;

    ['id', 'text', 'rootId'].forEach(prop => {
      Object.defineProperty(this, prop, {
        get: () => this.item[prop],
      });
    });
  }

  get children() {
    return Node.nodes.filter(item => item.rootId === this.id);
  }

  get isOpen() {
    return this.open;
  }

  set isOpen(value) {
    this.open = value;
  }

  static setNodes(items) {
    Node.nodes = items.map(item => {
      const isLeaf = items.some(i => i.rootId === item.id) === false;

      return new Node(item, isLeaf);
    });
  }
}
