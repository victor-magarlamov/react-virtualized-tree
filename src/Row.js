import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

export default class Row extends Component {
  static V_LINE_CLASS = 'rvtree_node_v-line';

  static H_LINE_CLASS = 'rvtree_node_h-line';

  static EXPAND_CLASS = 'rvtree_node_expand';

  static OPEN_CLASS = 'open';

  static CLOSE_CLASS = 'close';

  get lines() {
    const { height } = this.props.style;
    const { level, id, isLastNode, isLastBranch } = this.props.item;

    let numOfFullLines = level - 2;

    let levels = [];

    if (numOfFullLines > 0) {
      levels = Array.from(new Array(numOfFullLines).keys());
    }

    const vLineStyle = {
      height: `${height}px`,
    };

    const firstVLineStyle = { ...vLineStyle };
    const lastVLineStyle = { ...vLineStyle };
    if (isLastBranch) {
      firstVLineStyle.background = 'none';
    }

    if (isLastNode) {
      lastVLineStyle.height = `${height / 2}px`;
    }

    return (
      <Fragment>
        {level > 1 && (
          <div className={Row.V_LINE_CLASS} style={firstVLineStyle} />
        )}

        {levels.map(i => (
          <div
            key={`${id}-${i}`}
            className={Row.V_LINE_CLASS}
            style={vLineStyle}
          />
        ))}

        <div className={Row.V_LINE_CLASS} style={lastVLineStyle} />

        <div className={Row.H_LINE_CLASS} />
      </Fragment>
    );
  }

  get expandButton() {
    const { isLeaf, isOpen } = this.props.item;

    if (isLeaf) {
      return null;
    }

    const state = isOpen ? Row.OPEN_CLASS : Row.CLOSE_CLASS;

    return (
      <div
        className={`${Row.EXPAND_CLASS} ${Row.EXPAND_CLASS}-${state}`}
        onClick={this.handleExpand}
      />
    );
  }

  get checkbox() {
    const { onCheck, item } = this.props;

    if (!onCheck) {
      return null;
    }

    return (
      <input
        className="rvtree_node_checkbox"
        type="checkbox"
        name={`item_${item.id}`}
        value={false}
        onChange={this.handleCheckboxChanged}
      />
    );
  }

  get label() {
    const { item } = this.props;

    return <div onClick={this.handleClick}>{item.text}</div>;
  }

  handleExpand = () => {
    const { item, index, onExpand } = this.props;

    onExpand(item, index);
  };

  handleClick = () => {
    const { item, onClick } = this.props;

    if (!onClick) return;

    onClick(item.id);
  };

  handleCheckboxChanged = e => {
    const { item, onCheck } = this.props;

    onCheck(item.id, e.target.checked);
  };

  render() {
    const { style } = this.props;

    return (
      <div className="rvtree_node" style={style}>
        {this.lines}
        {this.expandButton}
        {this.checkbox}
        {this.label}
      </div>
    );
  }
}

Row.propTypes = {
  style: PropTypes.object.isRequired,
  onExpand: PropTypes.func.isRequired,
  index: PropTypes.number.isRequired,
  onClick: PropTypes.func,
  onCheck: PropTypes.func,
  item: PropTypes.shape({
    id: PropTypes.number,
    text: PropTypes.string,
    level: PropTypes.number,
    isLeaf: PropTypes.bool,
    isOpen: PropTypes.bool,
    isLastNode: PropTypes.bool,
    isLastBranch: PropTypes.bool,
    isRoot: PropTypes.bool,
  }).isRequired,
};
