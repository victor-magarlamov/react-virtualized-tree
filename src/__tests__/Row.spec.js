import React from 'react';
import Row from '../Row';
import Node from '../Node';
import renderer from 'react-test-renderer';
  
const item = {
  id: 1,
  text: 'text',
  rootId: null,
};

const props = {
  style: {
    height: 24,
  },
  onClick: () => {},
  onExpand: () => {},
  index: 0,
};

const rootNode = new Node(item, false);
const leafNode = new Node(item, true);

describe('Root node', () => {
  describe('when closed', () => {
    const component = renderer.create(
      <Row {...props} item={rootNode} />
    );

    let row = component.toJSON();

    it('has "rvtree_node_expand-close" class', () => {
      expect(row).toMatchSnapshot();
    });
  });
  
  describe('when opened', () => {
    rootNode.isOpen = true;

    const component = renderer.create(
      <Row {...props} item={rootNode} />
    );

    let row = component.toJSON();

    it('has "rvtree_node_expand-open" class', () => {
      expect(row).toMatchSnapshot();
    });
  });
});

describe('Leaf node', () => {
  const component = renderer.create(
    <Row {...props} item={leafNode} />
  );

  let row = component.toJSON();

  it('has not expand button', () => {
    expect(row).toMatchSnapshot();
  });
});
