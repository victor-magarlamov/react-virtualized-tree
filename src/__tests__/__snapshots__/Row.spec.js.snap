// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaf node has not expand button 1`] = `
<div
  className="rvtree_node"
  style={
    Object {
      "height": 24,
    }
  }
>
  <div
    className="rvtree_node_v-line"
    style={
      Object {
        "height": "24px",
      }
    }
  />
  <div
    className="rvtree_node_h-line"
  />
  <div
    onClick={[Function]}
  >
    text
  </div>
</div>
`;

exports[`Root node when checkable has checkbox 1`] = `
<div
  className="rvtree_node"
  style={
    Object {
      "height": 24,
    }
  }
>
  <div
    className="rvtree_node_v-line"
    style={
      Object {
        "height": "24px",
      }
    }
  />
  <div
    className="rvtree_node_h-line"
  />
  <div
    className="rvtree_node_expand rvtree_node_expand-open"
    onClick={[Function]}
  />
  <div
    onClick={[Function]}
  >
    text
  </div>
</div>
`;

exports[`Root node when closed has "rvtree_node_expand-close" class 1`] = `
<div
  className="rvtree_node"
  style={
    Object {
      "height": 24,
    }
  }
>
  <div
    className="rvtree_node_v-line"
    style={
      Object {
        "height": "24px",
      }
    }
  />
  <div
    className="rvtree_node_h-line"
  />
  <div
    className="rvtree_node_expand rvtree_node_expand-close"
    onClick={[Function]}
  />
  <div
    onClick={[Function]}
  >
    text
  </div>
</div>
`;

exports[`Root node when opened has "rvtree_node_expand-open" class 1`] = `
<div
  className="rvtree_node"
  style={
    Object {
      "height": 24,
    }
  }
>
  <div
    className="rvtree_node_v-line"
    style={
      Object {
        "height": "24px",
      }
    }
  />
  <div
    className="rvtree_node_h-line"
  />
  <div
    className="rvtree_node_expand rvtree_node_expand-open"
    onClick={[Function]}
  />
  <div
    onClick={[Function]}
  >
    text
  </div>
</div>
`;
